src/notification/notification.service.ts

```diff
@@ -0,0 +1,66 @@
+import { BadRequestException, Injectable } from '@nestjs/common';
+import * as firestoreAdmin from 'firebase-admin/firestore';
+import { messaging } from 'firebase-admin';
+import { Messaging } from 'firebase-admin/lib/messaging/messaging';
+import { Device, NotificationMessage } from './interface/notification.interface';
+import { Message, MulticastMessage } from 'firebase-admin/lib/messaging/messaging-api';
+@Injectable()
+export class NotificationService {
+  private collection: firestoreAdmin.CollectionReference;
+  private messenger: Messaging;
+  constructor() {
+    this.collection = firestoreAdmin.getFirestore().collection('aegis-notification');
+    this.messenger = messaging();
+  }
+  async validateRegistrationToken(registrationToken: string): Promise<boolean> {
+    try {
+      return !!(await this.messenger.send({ token: registrationToken }, true));
+    } catch (err) {
+      return false;
+    }
+  }
+  async registerDevice(uid: number, registrationToken: string): Promise<boolean> {
+    if (!this.validateRegistrationToken(registrationToken)) throw new BadRequestException('Invalid token');
+    const deviceRef = this.collection.doc(uid.toString());
+    const res = await deviceRef.set({
+      token: registrationToken,
+      timestamp: Date.now(),
+    });
+    return !!res;
+  }
+  async findDeviceByUid(uid: number): Promise<Device> {
+    const ref = await this.collection.doc(uid.toString()).get();
+    return ref.data() as Device;
+  }
+  async findManyDevicesByUids(uids: number[]) {
+    const stringUids = uids.map((uid) => uid.toString());
+    const results = [];
+    for (let i = 0; i < stringUids.length; i += 10) {
+      const chunk = stringUids.slice(i, i + 10);
+      const ref = await this.collection.where(firestoreAdmin.FieldPath.documentId(), 'in', chunk).get();
+      Array.prototype.push.apply(results, ref.docs);
+    }
+    return results.map((doc) => doc.data()).filter((device) => !!device) as Device[];
```

Comment:
this too

Reply:
This is to filter out `undefined` value which is possible from `doc.data()`